### PR TITLE
[Merged by Bors] - refactor: redefine `Bundle.TotalSpace`

### PR DESCRIPTION
**Pull request by urkud**
Forward-port leanprover-community/mathlib#19221

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Bundle.lean [hunk 1]
```diff
@@ -17,179 +17,149 @@ should contain all possible results that do not involve any topology.
 
 We represent a bundle `E` over a base space `B` as a dependent type `E : B → Type _`.
 
-We provide a type synonym of `Σ x, E x` as `Bundle.TotalSpace E`, to be able to endow it with
-a topology which is not the disjoint union topology `Sigma.TopologicalSpace`. In general, the
-constructions of fiber bundles we will make will be of this form.
+We define `Bundle.TotalSpace F E` to be the type of pairs `⟨b, x⟩`, where `b : B` and `x : E x`.
```
Comment (negiizhao) [inline Mathlib/Data/Bundle.lean:20]:
```suggestion
We define `Bundle.TotalSpace F E` to be the type of pairs `⟨b, x⟩`, where `b : B` and `x : E b`.
```

File: Mathlib/Topology/VectorBundle/Basic.lean [hunk 2]
```diff
@@ -639,16 +639,15 @@ instance moduleFiber (x : B) : Module R (Z.Fiber x) :=
...
 @[reducible, simp, mfld_simps]
-protected def proj : TotalSpace Z.Fiber → B :=
+protected def proj : TotalSpace F Z.Fiber → B :=
   TotalSpace.proj
 #align vector_bundle_core.proj VectorBundleCore.proj
 
 /-- The total space of the vector bundle, as a convenience function for dot notation.
-It is by definition equal to `Bundle.TotalSpace Z.fiber`, a.k.a. `Σ x, Z.fiber x` but with a
-different name for typeclass inference. -/
+It is by definition equal to `Bundle.TotalSpace Z.fiber`. -/
```
Comment (negiizhao) [inline Mathlib/Topology/VectorBundle/Basic.lean:647]:
```suggestion
It is by definition equal to `Bundle.TotalSpace F Z.Fiber`. -/
```

File: Mathlib/Data/Bundle.lean [hunk 3]
```diff
@@ -17,179 +17,149 @@ should contain all possible results that do not involve any topology.
...
+open Function Set
 
 namespace Bundle
 
-variable {B : Type _} (E : B → Type _)
+variable {B F : Type _} (E : B → Type _)
 
-/-- `Bundle.TotalSpace E` is the total space of the bundle `Σ x, E x`.
-This type synonym is used to avoid conflicts with general sigma types.
+/-- `Bundle.TotalSpace E` is the total space of the bundle. It consists of pairs
```
Comment (negiizhao) [inline Mathlib/Data/Bundle.lean:51]:
```suggestion
/-- `Bundle.TotalSpace F E` is the total space of the bundle. It consists of pairs
```

File: Mathlib/Data/Bundle.lean [hunk 4]
```diff
@@ -17,179 +17,149 @@ should contain all possible results that do not involve any topology.
...
+`(proj : B, snd : E proj)`.
 -/
-def TotalSpace :=
-  Σx, E x
+@[ext]
+structure TotalSpace (F : Type _) (E : B → Type _) where
+  /-- `bundle.total_space.proj` is the canonical projection `Bundle.TotalSpace E → B` from the
```
Comment (negiizhao) [inline Mathlib/Data/Bundle.lean:56]:
```suggestion
  /-- `Bundle.TotalSpace.proj` is the canonical projection `Bundle.TotalSpace F E → B` from the
```

File: Mathlib/Data/Bundle.lean [hunk 5]
```diff
@@ -17,179 +17,149 @@ should contain all possible results that do not involve any topology.
...
--- mathport name: «expr *ᵖ »
 @[inherit_doc]
-notation f " *ᵖ " E => Pullback f E
+notation f " *ᵖ " E:arg => Pullback f E
+
+instance {f : B' → B} {x : B'} [Nonempty (E (f x))] : Nonempty ((f *ᵖ E) x) :=
+  ‹Nonempty (E (f x))›
 
-/-- Natural embedding of the total space of `f *ᵖ E` into `B' × TotalSpace E`. -/
+/-- Natural embedding of the total space of `f *ᵖ E` into `B' × total_space E`. -/
```
Comment (negiizhao) [inline Mathlib/Data/Bundle.lean:142]:
```suggestion
/-- Natural embedding of the total space of `f *ᵖ E` into `B' × TotalSpace E`. -/
```

File: Mathlib/Topology/FiberBundle/Basic.lean [hunk 6]
```diff
@@ -17,11 +17,9 @@ Mathematically, a (topological) fiber bundle with fiber `F` over a base `B` is a
...
-In our formalism, a fiber bundle is by definition the type
-`Bundle.TotalSpace E` where `E : B → Type*` is a function associating to
-`x : B` the fiber over `x`. This type `Bundle.TotalSpace E` is just a type synonym for
-`Σ (x : B), E x`, with the interest that one can put another topology than on `Σ (x : B), E x`
-which has the disjoint union topology.
+In our formalism, a fiber bundle is by definition the type `Bundle.TotalSpace F E` where
+`E : B → Type _` is a function associating to `x : B` the fiber over `x`. This type
+`Bundle.TotalSpace F E` is a type of pairs `⟨proj : B, snd : E proj⟩`.
 
 To have a fiber bundle structure on `Bundle.TotalSpace E`, one should
```
Comment (negiizhao) [inline Mathlib/Topology/FiberBundle/Basic.lean:24]:
```suggestion
To have a fiber bundle structure on `Bundle.TotalSpace F E`, one should
```

and also the 4th line below

File: Mathlib/Topology/FiberBundle/Basic.lean [hunk 7]
```diff
@@ -54,19 +52,18 @@ fiber bundle from trivializations given as local equivalences with minimum addit
...
 * `FiberBundleCore ι B F` : structure registering how changes of coordinates act
   on the fiber `F` above open subsets of `B`, where local trivializations are indexed by `ι`.
 
 Let `Z : FiberBundleCore ι B F`. Then we define
 
 * `Z.Fiber x`     : the fiber above `x`, homeomorphic to `F` (and defeq to `F` as a type).
-* `Z.TotalSpace` : the total space of `Z`, defined as a `Type` as `Σ (b : B), F`, but with a
-  twisted topology coming from the fiber bundle structure. It is (reducibly) the same as
-  `Bundle.TotalSpace Z.Fiber`.
+* `Z.TotalSpace`  : the total space of `Z`, defined as `Bundle.TotalSpace F Z.fiber` with a custom
```
Comment (negiizhao) [inline Mathlib/Topology/FiberBundle/Basic.lean:63]:
```suggestion
* `Z.TotalSpace`  : the total space of `Z`, defined as `Bundle.TotalSpace F Z.Fiber` with a custom
```

File: Mathlib/Topology/FiberBundle/Basic.lean [hunk 8]
```diff
@@ -440,10 +435,9 @@ instance topologicalSpaceFiber (x : B) : TopologicalSpace (Z.Fiber x) := ‹_›
 #align fiber_bundle_core.topological_space_fiber FiberBundleCore.topologicalSpaceFiber
 
 /-- The total space of the fiber bundle, as a convenience function for dot notation.
-It is by definition equal to `Bundle.TotalSpace Z.Fiber`, a.k.a. `Σ x, Z.Fiber x` but with a
-different name for typeclass inference. -/
+It is by definition equal to `Bundle.TotalSpace Z.fiber`. -/
```
Comment (negiizhao) [inline Mathlib/Topology/FiberBundle/Basic.lean:438]:
```suggestion
It is by definition equal to `Bundle.TotalSpace F Z.Fiber`. -/
```

File: Mathlib/Topology/FiberBundle/Trivialization.lean [hunk 9]
```diff
@@ -250,52 +252,52 @@ variable [∀ x, Zero (E x)]
 
 /-- A fiberwise inverse to `e`. This is the function `F → E b` that induces a local inverse
 `B × F → total_space E` of `e` on `e.baseSet`. It is defined to be `0` outside `e.baseSet`. -/
```
Comment (negiizhao) [inline Mathlib/Topology/FiberBundle/Trivialization.lean:254]:
```suggestion
`B × F → TotalSpace E` of `e` on `e.baseSet`. It is defined to be `0` outside `e.baseSet`. -/
```

File: Mathlib/Topology/FiberBundle/Trivialization.lean [hunk 10]
```diff
@@ -612,44 +614,44 @@ variable [∀ x, Zero (E x)]
 
 /-- A fiberwise inverse to `e'`. The function `F → E x` that induces a local inverse
 `B × F → total_space E` of `e'` on `e'.baseSet`. It is defined to be `0` outside `e'.baseSet`. -/
```
Comment (negiizhao) [inline Mathlib/Topology/FiberBundle/Trivialization.lean:616]:
```suggestion
`B × F → TotalSpace E` of `e'` on `e'.baseSet`. It is defined to be `0` outside `e'.baseSet`. -/
```

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**